src/mixins/text_style.mixin.js

```diff
@@ -319,6 +319,86 @@
      */
     _deleteLineStyle: function(lineIndex) {
       delete this.styles[lineIndex];
+    },
+
+    /**
+     * Returns the array form of an object's styles property where styles are grouped in ranges
+     * rather than per character. This format is less verbose, and is better suited for storage
+     * so it is used in serialization (not during runtime).
+     * @return {Array}
+     */
+    _stylesToArray: function() {
+      // clone style structure to prevent mutation
+      var styles = fabric.util.object.clone(this.styles, true),
+          textLines = this.text.split('\n'),
+          charIndex = -1, prevStyle = {}, stylesArray = [];
+      //loop through each textLine
+      for (var i = 0; i < textLines.length; i++) {
+        if (!styles[i]) {
+          //no styles exist for this line, so add the line's length to the charIndex total
+          charIndex += textLines[i].length;
+          continue;
+        }
+        //loop through each character of the current line
+        for (var c = 0; c < textLines[i].length; c++) {
+          charIndex++;
+          var thisStyle = styles[i][c];
+          //check if style exists for this character
+          if (thisStyle) {
+            //need to call _hasStyleChangedForSvg, not _hasStyleChanged so that overline,
+            //underline, and line-through are also checked
+            var styleChanged = this._hasStyleChangedForSvg(prevStyle, thisStyle);
+            if (styleChanged) {
+              stylesArray.push({
+                start: charIndex,
+                end: charIndex + 1,
+                style: thisStyle
+              });
+            }
+            else {
+              //if style is the same as previous character, increase end index
+              stylesArray[stylesArray.length - 1].end++;
+            }
+          }
+          prevStyle = thisStyle || {};
+        }
+      }
+      return stylesArray;
+    },
+
+    /**
+     * Returns the object form of the styles property where styles are assigned per
+     * character rather than grouped by range. This format is more verbose, and is
+     * only used during runtime (not for serialization/storage)
+     * @return {Object}
+     */
+    _stylesFromArray: function() {
+      if (!Array.isArray(this.styles)) {
+        return this.styles;
+      }
+      var textLines = this.text.split('\n'),
```

Comment:
Does this need to happen before text is parsed or can it happen right after? 
Saving some computation effort

Reply:
Hmm...not sure I follow. Where are you suggesting `text.split('\n')` could go instead?